clang/include/clang/Basic/Attr.td

```diff
@@ -1135,6 +1135,20 @@ def SYCLKernel : InheritableAttr {
   let Documentation = [SYCLKernelDocs];
 }
 
+def SYCLSpecialClass: InheritableAttr {
+  let Spellings = [Clang<"sycl_special_class">];
+  let Subjects = SubjectList<[CXXRecord]>;
+  let LangOpts = [SYCLIsDevice, SYCLIsHost];
+  // TODO: Add doc
+  let Documentation = [Undocumented];
+  let Args = [
+    EnumArgument<"SpecialClassKind", "SpecialClassKind",
+                 [ "accessor", "sampler", "stream", "" ],
+                 [ "Accessor", "Sampler", "Stream", "Generic" ], 1>
```

Comment:
I have mainly looked at the attribute definition here.

This seems to me to be very specific to SYCL and does not cover every special types (like specialization constant). Also if you want to add a new one, you need to change the compiler.

One of the idea I had was to make the attribute sensitive to `__init` and `__finalize` (if present) and maintain the derivation of the kernel arguments from the `__init` argument list. This allow to add new special types without the need to touch the compiler and you are actually SYCL agnostic. As this is a special class, the runtime should be able to query what the class is and act accordingly. I never had the time to fully explore and test the idea, so it might be flawed. There is also the case of the specialization constant, where you may or may not want to pass the value as an argument, but maybe this could simply be handled using macros.

One drawback I see from a "generic attribute" point of view, this is C++ specific as `__init` and `__finalize` are expected to be member functions. But the functions could be specify as part of the argument.

> I think this might be OpenCL/SPIR-V specific attribute instead of SYCL specific attribute

I disagree, OpenCL doesn't need to apply any kind of processing to build the entry point. SPIR-V is neutral to that regard as well.

This is to allow regularization to an underlying programming model. So OpenCL or CUDA at the moment, but we could very well imagine something OpenMP or Vulkan compatible.

Reply:
> This seems to me to be very specific to SYCL and does not cover every special types (like specialization constant). Also if you want to add a new one, you need to change the compiler.

Yes, it doesn't cover specialization constant, I marked it as unfinished TODO in the PR description. In case of new special type... If the new special type needs some special handling that differs from others, you need to change the compiler anyway, otherwise "generic" value of the attribute can be used.

> One of the idea I had was to make the attribute sensitive to __init and __finalize (if present) and maintain the derivation of the kernel arguments from the __init argument list.

Could you please describe how it would look like? I'm not sure that I got the idea.

> As this is a special class, the runtime should be able to query what the class is and act accordingly.

With the current implementation the runtime takes the address of kernel object and uses offsets encoded by the integration header. It also uses information about kernel object field types to cast area of memory accessed by using of address and offset to proper type and perform handling, it cannot query the type from raw memory. In this case we probably could teach integration header to emit information that some kernel argument is a 'special class' and add some base class for each class that is handled specially by the compiler, so this base class will hold kind of special class and it will be possible to query it, I guess. But I'm not really sure since I don't have a lot of experience with runtime.

Reply:
>  If the new special type needs some special handling that differs from others, you need to change the compiler anyway, otherwise "generic" value of the attribute can be used.

This attribute doesn't do much. Special types processing boils down to fully or partially breakdown the structure into valid fields for your programming model. The rest is the associated processing in the runtime.

> Could you please describe how it would look like? I'm not sure that I got the idea.

I'll take the `accessor` as an example.
ATM (and last time I checked) the processing of the special type `sycl::accessor` is roughly done as follow:

- Check if sycl special type `sycl::accessor`, if yes
- Look up for the member function `sycl::accessor::__init`
- Copy the `sycl::accessor::__init` arguments into the kernel function argument list
- Allocate a `sycl::accessor` object and call `sycl::accessor::__init` with the kernel argument

This is spread out in different places, but that's the basic idea. The important thing to notice is if you replace `sycl::accessor` by `sycl::sampler`, `sycl::stream` or other, the processing remains the same. The type `sycl::stream` has an extra step which is to call `__finalize` at the end of the kernel.

So in a more generic way, you could have `__attribute__((sycl_special_class))` (leaving the renaming exercise for later) that actually trigger the process above in a more generic way:

- Check if `<typename>` has attribute `sycl_special_class`, if yes
- Look up for the member function `<typename>::__init`
- Add kernel function has friend of the class
- Copy the `<typename>::__init` arguments into the kernel function argument list
- Allocate a `<typename>` object and call `<typename>::__init` with the kernel argument
- Look up for the member function `<typename>::__finalize`, if does have the function
- Call `<typename>::__finalize` at the end of the kernel

So for example:
```cpp
class __attribute__((sycl_special_class)) MySpecialType {
  int Field1;
  int Field2;
  void __init(int F1) {
    Field1 = F1;
    Field2 = F1;
  }
  void __finalize() {}
public:
  MySpecialType() = default;
  int getF2() const { return Field2; }
};
```
If used in a kernel argument
```
MySpecialType T;
cgh.single_task([=]() {
  T.getF2();
});
```
This would trigger the following kernel entry point in the AST:

```cpp
void __sycl_kernel(int F1) {
  MySpecialType T;
  T.__init(F1);
  // finish rebuilding the lambda + call
  T.__finalize()
}
```

> With the current implementation the runtime takes the address of kernel object and uses offsets encoded by the integration header. It also uses information about kernel object field types to cast area of memory accessed by using of address and offset to proper type and perform handling, it cannot query the type from raw memory. In this case we probably could teach integration header to emit information that some kernel argument is a 'special class' and add some base class for each class that is handled specially by the compiler, so this base class will hold kind of special class and it will be possible to query it, I guess. But I'm not really sure since I don't have a lot of experience with runtime.

So the integration header already emits what kind of fields it is processing. Changing to this would only require the capability to query it in the runtime. Since the std-layout requirement was lifted, a simple base class and reinterpret cast should do the job.

Another approach is to make the attribute take an ID as argument (the ID is chosen by the runtime)  and forward it in the integration header (so in place of the field currently emitted). The runtime then only have to read-it directly. Either way, the compiler remains agnostic to the field it is processing.


Hopes this make the idea clearer.

Reply:
Yes, it definitely makes the idea clearer. Thanks for so detailed explanation!
I'll give it a try. It will require more changes across front-end and runtime library, so it will take some time. I will update this PR once I have something ready.